Add documentation on plan life cycle and instance admission controller (#226)
kudobuilder/www: content/docs/what-is-kudo.md

    trigger: "update"
```

> Each parameter has a `trigger` field which specifies what plan will be executed when this parameter changes. The operations required to safely update a running application can vary depending on which parameter is being updated. For instance, updating the `NODE_COUNT` may require a simple update of the deployment, whereas updating the `NODE_MEM_MIB` may require rolling out a new version via a canary or blue/green deployment. If no trigger is specified, the default `deploy` plan will be executed. You can also trigger plans manually, without the corresponding parameter update. See [plans documentation](developing-operators/plans.md#executing-plans) for more details.

Reviewer: ```suggestion
Each parameter has a `trigger` field which specifies what plan will be executed when this parameter changes. The operations required to safely update a running application can vary depending on which parameter is being updated. For instance, updating the `NODE_COUNT` may require a simple update of the deployment, whereas updating the `NODE_MEM_MIB` may require rolling out a new version via a canary or blue/green deployment. If no trigger is specified, the default `deploy` plan will be executed.

 You can also trigger plans manually, without the corresponding parameter update. See [plans documentation](developing-operators/plans.md#executing-plans) for more details.
```

```bash
$ kubectl kudo install kafka --instance dev-kafka
```

> ## Executing Plans

Reviewer: Note that there already is an "Operator Plans" section above. It says that more on plans can be read in `/docs/developing-operators/plans.html` so adding more info about plans here is a bit of a hidden surprise. :-P

I think what you added actually belongs in that file not here, though it may make sense to add a few words to the above section just to mention that manual plan triggering is also possible.

Author: Agree. Done.

In general, KUDO manager will automatically [execute a plan](what-is-kudo.md#operator-parameters) when the corresponding parameter changes. However, sometimes this is not enough. Sometimes you need to execute a plan manually like running a periodic `backup` plan or executing `restore` plan in case of data corruption. Such plans typically do not need a corresponding parameter. KUDO [v0.11.0](https://github.com/kudobuilder/kudo/releases/tag/v0.11.0) introduced new feature: manual plan execution. In a nutshell, you can now:

```bash
$ kubectl kudo plan trigger --name deploy --instance my-instance
```

which will execute `deploy` plan on `my-instance`. While this looks relatively easy on the surface, the devil is in detail, so let's take a closer look.

### Plan Life Cycle

> Having the ability to trigger multiple plans on demand raises the question: what happens if two plans run concurrently? The answer is: it depends. Should two plans be completely independent of each other (e.g. `deploy` deploys the services and `monitor` deploys the monitoring pods) both can be executed in parallel. But if two plans in question are `backup` and `restore` or `deploy` and `migrate`? Some plans are incompatible with others. A few may not even be compatible with themselves. While it is probably ok to restart a running `deploy` plan, a `restore` plan might not be reentrant because of possible data corruption.

Reviewer: ```suggestion
Having the ability to trigger multiple plans on demand raises the question: what happens if two plans run concurrently? The answer is: it depends. Should two plans be completely independent of each other (e.g. `deploy` deploys the services and `monitor` deploys the monitoring pods) both can be executed in parallel. But if two plans in question are `backup` and `restore` or `deploy` and `migrate`? Some plans are incompatible with others. A few may not even be compatible with another running instance of themselves. While it is probably ok to restart a running `deploy` plan, a `restore` plan might not be reentrant because of possible data corruption.
```

Author: I reworded it differently.

We're planning to explore the realm of plan compatibility further in the future. Annotating plan affinity and anti-affinity, reentrant vs non-reentrant plans, plan cancelation and transient plan parameters are some of the topics we're examining. All contributions and feedback are highly welcome.

Having all this in mind how can we ensure correct plan execution? Meet Kubernetes admission controllers.

### Admission Controllers

> In a nutshell, Kubernetes admission controllers are plugins that govern and enforce how the cluster is used. They can be thought of as a gatekeeper that intercepts (authenticated) API requests and may change the request object or deny the request altogether. Kubernetes already comes with a bunch of these [perinstalled](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) which govern everything from user authorization to the namespace life cycle.

Reviewer: ```suggestion
In a nutshell, Kubernetes admission controllers are plugins that govern and enforce how the cluster is used. They can be thought of as a gatekeeper that intercepts (authenticated) API requests and may change the request object or deny the request altogether. Kubernetes already comes with a bunch of these [pre-installed](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) which govern everything from user authorization to the namespace life cycle.
```

KUDO manager employs an Instance admission controller that governs changes to Instances, making sure that plans do not interfere. Schematically this looks like the following:
![Instance update](/images/admission-instance-update.png)

> Instance admission controller governs any update to the Instance either through manual plan execution or Instance parameter updates. The general rule of thumb is the following: a plan should be terminal (either successfully with status `COMPLETE` or unsuccessfully with `FATAL_ERROR`) before a new plan is allowed to start. A singular plan can be restarted so we assume all plans to be reentrant at least for now. While this might not be true for all plans, we think that it covers the 80/20 case e.g. when a `deploy` plan is stuck and must be restarted with less memory per node. In case a request is rejected, the Instance controller returns an error explaining why exactly the update was denied.

Reviewer: ```suggestion
Instance admission controller governs any update to the Instance either through manual plan execution or Instance parameter updates. The general rule of thumb is the following: all plans should be terminal (either successfully with status `COMPLETE` or unsuccessfully with `FATAL_ERROR`) before a new plan is allowed to start. A singular plan can be restarted so we assume all plans to be reentrant at least for now. While this might not be true for all plans, we think that it covers the 80/20 case e.g. when a `deploy` plan is stuck and must be restarted with less memory per node. In case a request is rejected, the Instance controller returns an error explaining why exactly the update was denied.
```

The admission controller would also reject parameter updates that would trigger multiple distinct plans. There are a few exceptions too: for example, a `cleanup` plan is special and is executed when an Instance is deleted. `cleanup` can not be executed manually and is allowed to override any existing plan (since the Instance is being deleted anyway).

> ### Installing Admission Controller

Reviewer: I think this belongs in /docs/#install-kudo-into-your-cluster, perhaps with a pointer from here.

Author: I agree that it doesn't really belong here. However, I don't see a dedicated KUDO installation page.  We have some [kudo init cli examples](https://github.com/kudobuilder/kudo.dev/blob/master/content/docs/cli.md#kudo-init) and a [kudo init admin runbook](https://github.com/kudobuilder/kudo.dev/blob/master/content/docs/runbooks/admin/initialize-kudo.md). Would you prefer a third doc? Or extending an existing one?

Reviewer: Well, on top of that we have the very short [section in quick start](https://github.com/kudobuilder/kudo.dev/blob/master/content/docs/README.md#install-kudo-into-your-cluster) (that I meant to point to above) which mostly overlaps with the runbook (but not completely, since the former also mentions `--dry-run` while the latter has verification instructions).

My preference would be:
- document `--webhook` in the `init` section of `cli.md`, and also mention the cert-manager dependency there
- in "Install KUDO into your cluster" section of README.md also mention the webhook among the other listed components, and link to `cli.md` for details
- for completeness, add a "further information" section to the runbook with a link to `cli.md`

Author: Done and done 👍  PTAL
